packages/ethereum-contracts/contracts/mocks/CFALibraryMock.sol

```diff
@@ -0,0 +1,400 @@
+                ))
+            )
+        );
+    }
+
+    function createFlowTest(
+        ISuperfluidToken token,
+        address receiver,
+        int96 flowRate
+    ) public {
+        cfaV1.createFlow(receiver, token, flowRate);
+    }
+
+    function createFlowWithUserDataTest(
+        ISuperfluidToken token,
+        address receiver,
+        int96 flowRate,
+        bytes memory userData
+    ) public {
+        cfaV1.createFlow(receiver, token, flowRate, userData);
+    }
+
+    function updateFlowTest(
+        ISuperfluidToken token,
+        address receiver,
+        int96 flowRate
+    ) public {
+        cfaV1.updateFlow(receiver, token, flowRate);
+    }
+
+    function updateFlowWithUserDataTest(
+        ISuperfluidToken token,
+        address receiver,
+        int96 flowRate,
+        bytes memory userData
+    ) public {
+        cfaV1.updateFlow(receiver, token, flowRate, userData);
+    }
+
+    function deleteFlowTest(
+        ISuperfluidToken token,
+        address receiver
+    ) public {
+        cfaV1.deleteFlow(address(this), receiver, token);
+    }
+
+    function deleteFlowWithUserDataTest(
+        ISuperfluidToken token,
+        address receiver,
+        bytes memory userData
+    ) public {
+        cfaV1.deleteFlow(address(this), receiver, token, userData);
+    }
+
+    function createFlowWithCtxTest(
+        bytes memory ctx,
+        address receiver,
+        ISuperfluidToken token,
+        int96 flowRate
+    ) public {
+        cfaV1.createFlowWithCtx(ctx, receiver,token, flowRate);
+    }
+
+    function createFlowWithCtxUserDataTest(
+        bytes memory ctx,
+        address receiver,
+        ISuperfluidToken token,
+        int96 flowRate,
+        bytes memory userData
+    ) public {
+        cfaV1.createFlowWithCtx(ctx, receiver, token, flowRate, userData);
+    }
+
+    function updateFlowWithCtxTest(
+        bytes memory ctx,
+        address receiver,
+        ISuperfluidToken token,
+        int96 flowRate
+    ) public {
+        cfaV1.updateFlowWithCtx(ctx, receiver, token, flowRate);
+    }
+
+    function updateFlowWithCtxUserDataTest(
+        bytes memory ctx,
+        address receiver,
+        ISuperfluidToken token,
+        int96 flowRate,
+        bytes memory userData
+    ) public {
+        cfaV1.updateFlowWithCtx(ctx, receiver, token, flowRate, userData);
+    }
+
+    function deleteFlowWithCtxTest(
+        bytes memory ctx,
+        address sender,
+        address receiver,
+        ISuperfluidToken token
+    ) public {
+        cfaV1.deleteFlowWithCtx(ctx, sender, receiver, token);
+    }
+
+    function deleteFlowWithCtxUserDataTest(
+        bytes memory ctx,
+        address sender,
+        address receiver,
+        ISuperfluidToken token,
+        bytes memory userData
+    ) public {
+        cfaV1.deleteFlowWithCtx(ctx, sender, receiver, token, userData);
+    }
+
+}
+
+contract RedirectAll is SuperAppBase {
```

Comment:
I understood that this is to reuse the redirectall for testing all the positive cases, right?

Though, to reduce amount of code in the repository, would it be possible if we reuse MFA.sol and use the helper library in that instead?

Reply:
Yes that is correct.

I think so re: MFA.sol. The goal of using redirectall was to ensure that the WithCtx versions of each function work properly inside of a Super App. To achieve the same with the MFA.sol app, I believe I'll need to create a new version of the MultiFlowApp which uses the CFA Lib to make calls to the CFA instead of the callAgreementWithContext(...) pattern. 

I can create a new version in contracts/test called MultiFlowAppLib.so which only changes the way that flows are created, updated, and deleted. Then I can reuse the tests written in ConstantFlowAgreement.test.js to ensure coverage. Does that work?

Reply:
Okay, I see.

Using the new CFAv1Library in MultiFlowApp.sol can be a separate task for later then.

And in order to keep some consistency, I suggest we consider the following naming and file location changes:

- [ ] name to CFAv1Library, reason: it is a library for CFAv1
  - CFAv1Library.sol
  - CFAv1Library.test.js
- [ ] CFAv1Library.sol to be moved to contracts/apps/, reason:  SuperAppBase.sol is there, and it's a folder for helping to build your own super apps
  - [ ]  Fix .solcover.js, to not to skip everything from apps/ (skip apps/SuperAppBase.sol only)
- [ ] CFALibraryMock.sol, and merge CFALibraryMock, RedirectAll and TradeableCashflowMock into one CFAv1LibraryTester contract. Reason: to keep the contract strictly for "testing" purpose.

Reply:
I will change the names 👍🏻

**On merging the mocks into CFAv1LibraryTester - the current test suite works as follows inside of CFALibrary.test.js:**
1) CFALibraryMock.sol includes a list of functions for creating, updating, and deleting flows which are implemented using the CFAv1Library.

2) The first set of tests call the functions on the CFALibraryMock contract to ensure that each function opens up streams properly to another EOA.

3) The second and third set of tests also call functions on the CFALibraryMock contract, but this time they set the recipient of each flow to be the deployed TradeableCashflowMock contract's address.
Because the callbacks inside of RedirectAll.sol use the WithCtx versions of the create, update, and delete functions of the CFAv1Library, we can check whether they work as expected when a stream is created, updated, or deleted that includes the Tradeable cashflow as the receiver. To do so, we can check to see if the owner of the Tradeable Cashflow contract is receiving a flow equal to the flow being sent into the contract. 

4) Each flow that is sent from the CFALibraryMock contract to the Tradeable Cashflow contract makes use of the overloaded version of the create, update, and delete functions that takes `userData` as a final parameter. The `userData` value is then stored to a public variable inside of the associated callback in the redirectAll contract. The test suite will then check that the userData was properly included.

If I combine the two into the same contract, I will need to reformat the tests somewhat in order to maintain test coverage on the functions which include userData.

Is there another way I could check to see that userData is being passed to the overloaded functions properly? Maybe I could check w Vincent to figure this one out.